ckanext/canada/cli.py

```diff
@@ -699,81 +800,140 @@ def _trim_package(pkg):
+    failure_reason = ''
+    failure_trace = ''
     try:
         portal.call_action('datastore_search', {'resource_id': resource['id'], 'limit': 0})
-        if t_hash.get(resource['id']) \
-                and t_hash.get(resource['id']) == resource.get('hash')\
+        if resource_file_hashes.get(resource['id']) \
+                and resource_file_hashes.get(resource['id']) == resource.get('hash')\
                 and _datastore_dictionary(portal, resource['id']) == resource_details['data_dict']:
             if verbose:
                 action += '\n  File hash and Data Dictionary has not changed, skipping DataStore for %s...' % resource['id']
             return action
         else:
-            portal.call_action('datastore_delete', {"id": resource['id'], "force": True})
-            action += '\n  datastore-deleted for ' + resource['id']
+            try:
+                portal.call_action('datastore_delete', {"id": resource['id'], "force": True})
+                action += '\n  datastore-deleted for ' + resource['id']
+            except Exception as e:
+                failure_reason = 'datastore_delete[resource_id=%s]' % resource['id']
+                failure_trace = traceback.format_exc()
+                error += '\n  datastore-create-failed for %s %s' % (resource['id'], str(e))
+                if verbose:
+                    error += '\n    Failed with Error: %s' % failure_trace
+                pass
     except NotFound:
         # not an issue, resource does not exist in datastore
         if verbose:
             action += '\n  DataStore does not exist for resource %s...trying to create it...' % resource['id']
         pass
 
-    portal.call_action('datastore_create',
-                       {"resource_id": resource['id'],
-                        "fields": resource_details['data_dict'],
-                        "force": True})
-
-    action += '\n  datastore-created for ' + resource['id']
+    datastore_created = False
+    try:
+        portal.call_action('datastore_create',
+                           {"resource_id": resource['id'],
+                            "fields": resource_details['data_dict'],
+                            "force": True})
+
+        action += '\n  datastore-created for ' + resource['id']
+        datastore_created = True
+    except Exception as e:
+        _failure_reason = 'datastore_create[resource_id=%s]' % resource['id']
+        _failure_trace = traceback.format_exc()
+        if failure_reason:
+            failure_reason += ',%s' % _failure_reason  # comma separate multiple failure reasons
+            failure_trace += '\n\n%s' % _failure_trace  # separate multiple failure traces with newlines
+        else:
+            failure_reason = _failure_reason
+            failure_trace = _failure_trace
+        error += '\n  datastore-create-failed for %s %s' % (resource['id'], str(e))
+        if verbose:
+            error += '\n    Failed with Error: %s' % failure_trace
+        pass
```

Comment:
These try: except: blocks are repeated a number of times. You could create e.g. a `capture_exception_details(trying_to)` context manager to reduce the duplication and these blocks could look like:
```python
with capture_exception_details('datastore-create'):
    portal.call_action('datastore_create',
                   {"resource_id": resource['id'],
                    "fields": resource_details['data_dict'],
                    "force": True})
    action += '\n  datastore-created for ' + resource['id']
    datastore_created = True
```

Reply:
@wardi I added a context manager class which handles the exceptions on exit and formats them into class variables. They are then used across the code to do the remainder of the formatting/logic for the `copy-dataset` exceptions. Because we do not want to blanket all of the action calls and allow for as much of the package to get synced as possible, there is still some code reuse in the methods as we need to keep the loops to allow for this.

At least it is a bit better now, unless if there is an even better way to store the formatted exceptions, reasons, and errors in an object? I will keep trying to work on that so I can minimize the code even further

Reply:
yeee I cannot really think of a way to do this. I think it is fine though? this way it would allow us to do other things in conditional statements in the future